docs - countif (#29104)
metabase/metabase: docs/questions/query-builder/expressions/countif.md

---
title: CountIf
---

# CountIf

> `CountIf` counts the total (not unique) rows in a table that match a condition.

Reviewer: ```suggestion
`CountIf` counts the total number of rows in a table that match a condition. `CountIf` counts every row, not just unique rows, so rows with duplicate values still count toward the total.
```

Maybe this construction? It's a little wordier, but maybe easier to parse.

Syntax: `CountIf(condition)`.

Example: in the table below, `CountIf([Plan] = "Basic")` would return 3.

| ID  | Plan        |
|-----|-------------|
| 1   | Basic       |
| 2   | Basic       |
| 3   | Basic       |
| 4   | Business    |
| 5   | Premium     |

> [Aggregations](../expressions-list.md#aggregations) like `CountIf` should be added to the query builder's [**Summarize** menu](../../query-builder/introduction.md#summarizing-and-grouping-by) > **Custom Expression** (scroll down in the menu if needed).

## Parameters

> `condition` is a [function](../expressions-list.md#functions) or [conditional statement](../expressions.md#conditional-operators) that returns a boolean value (`true` or `false`).

Reviewer: This phrasing is strange. Makes it seem like the word `condition` is a function you can invoke (like `condition()`.

## Multiple conditions

| ID  | Plan        | Active Subscription |
|-----|-------------| --------------------|
| 1   | Basic       | true                |
| 2   | Basic       | true                | 
| 3   | Basic       | false               |
| 4   | Business    | false               |
| 5   | Premium     | true                |

> ### Mandatory matches

Reviewer: ```suggestion
### Required conditions
```
Multiple and mandatory are both abstract words that start with M. This makes them a bit easier to distinguish.

> To count the total rows in a table that match multiple mandatory conditions, combine the conditions using the `AND` operator:

Reviewer: ```suggestion
To count the total number of rows in a table that match multiple required conditions, combine the conditions using the `AND` operator:
```

```
CountIf(([Plan] = "Basic" AND [Active Subscription] = true))
```

This expression will return 2 on the sample data above (the total number of Basic plans that have an active subscription).

### Optional matches

To count the total rows in a table that match multiple optional conditions, combine the conditions using the `OR` operator:

```
CountIf(([Plan] = "Basic" OR [Active Subscription] = true))
```

> Returns 4 on the sample data.

Reviewer: ```suggestion
Returns 4 on the sample data: there are three Basic plans, plus one Premium plan has an active subscription.
```

Might as well spell it out.

### Some mandatory and some optional matches

To combine mandatory and optional conditions, group the conditions using parentheses:

```
CountIf(([Plan] = "Basic" OR [Plan] = "Business") AND [Active Subscription] = "false"))
```

> Returns 2 on the sample data.

Reviewer: ```suggestion
Returns 2 on the sample data: there are only two Basic or Business plans that lack an active subscription.
```

> Tip: make it a habit to put parentheses around your `AND` and `OR` groups to avoid making mandatory conditions optional (or vice versa).

## Conditional count by group

In general, to get a conditional count for a category or group, such as the number of inactive subscriptions per plan, you'll:

1. Write a `CountIf` expression with your conditions.
2. Add a [**Group by**](../../query-builder/introduction.md#summarizing-and-grouping-by) column in the query builder.

Using the sample data:

| ID  | Plan        | Active Subscription |
|-----|-------------| --------------------|
| 1   | Basic       | true                |
| 2   | Basic       | true                | 
| 3   | Basic       | false               |
| 4   | Business    | false               |
| 5   | Premium     | true                |

Count the total number of inactive subscriptions per plan:

```
CountIf([Active Subscription] = false)
```

Alternatively, if your **Active Subscription** column contains `null` (empty) values that represent inactive plans, you could use:

```

> {% raw %}CountIf([Payment], [Plan] != true){% endraw %}

Reviewer: Do you need the raw tags here? Doesn't hurt, but this expression doesn't contain any mustachioed variables.

```

> > The "not equal" operator `!=` should be written as "!=".

Reviewer: ```suggestion
> The "not equal" operator should be written as `!=`.
```

Author: Oh yeah! This bit is weird because I have to explain why the code sample uses ≠ 

The actual formula should use != (which is what you get if you copy/paste the code sample)

The raw tags were a vestigial effort to force the != to show up 😅 

<img width="684" alt="Screenshot 2023-03-10 at 5 06 19 PM" src="https://user-images.githubusercontent.com/22361741/224437155-4330c80c-2435-4d2f-8f86-8750261d200a.png">

```

### CumulativeCount

`CountIf` doesn't do running counts. You'll need to combine the [CumulativeCount](../expressions-list.md#cumulativecount) aggregation with the [`case`](./case.md) function.

If our sample data is a time series:

| ID  | Plan        | Active Subscription | Created Date     |
|-----|-------------| --------------------|------------------|
| 1   | Basic       | true                | October 1, 2020  |
| 2   | Basic       | true                | October 1, 2020  | 
| 3   | Basic       | false               | October 1, 2020  |
| 4   | Business    | false               | November 1, 2020 |
| 5   | Premium     | true                | November 1, 2020 |

And we want to get the running count of active plans like this:

| Created Date: Month | Total Active Plans to Date |
|---------------------|----------------------------|
| October 2022        | 2                          |
| November 2022       | 3                          |

Create an aggregation from **Summarize** > **Custom expression**:

```
CumulativeCount(case([Active Subscription] = true, [Plan]))
```

You'll also need to set the **Group by** column to "Created Date: Month".

### SQL

> When you run a question using the [query builder](https://www.metabase.com/glossary/query_builder), Metabase will convert your graphical query settings (filters, summaries, etc.) into a query, and run that query against your database to get your results.

Reviewer: ```suggestion
When you run a question using the [query builder](https://www.metabase.com/glossary/query_builder), Metabase will convert your query builder settings (filters, summaries, etc.) into a SQL query, and run that query against your database to get your results.
```
